### ISSUE TITLE
Regex "*" token only captures one match with RegEx.search_all()

### ISSUE THREAD

**Issue by thebluetropics** [godotengine/godot#73920]
### Godot version

4.0.beta10

### System information

Windows 10

### Issue description

RegEx in godot only captures one, exactly one, using search_all() on a regex that contains "*". I don't know how to explain it in words, but I put a code to reproduce the problem below.

Why am I consider this as a bug? the same regex, "a*", in godot has different result of what RegExr does.

### Steps to reproduce

```gdscript
var regex = RegEx.new()
regex.compile("a*")

var str = "a captain eats crew"
var results = []

for result in regex.search_all(str):
  results.append(result.get_string())

print(results) # got ["a"], expected to be ["a", "a", "a", "a"]
# RegExr.com also gives me ["a", "a", "a", "a"] and not ["a"]
```

### Minimal reproduction project

No need

**Comment by AThousandShips** [comment]
Beta 10 was two months ago, can you test this with an up to date version

**Comment by thebluetropics** [comment]
@AThousandShips sure, I'll test out on a newer version, rq.

**Comment by thebluetropics** [comment]
@AThousandShips yes, I tested with newer version, 4.0.rc4. And the bug is still there with the exact same code as above.

**Comment by AThousandShips** [comment]
I am not sure why this breaks but if you just search for "a" not "a*" it works, I suspect because the search pattern causes some confusion as it can be empty therefore causes some kind of issue with restarts or anchors, for example it breaks with the string " a captain eats crew", indicating that when you add a repeat that yields an empty result you get some weirdness like that.

**Comment by AThousandShips** [comment]
To clarify what happens:
In the example you gave it searches and finds "a", then it starts after "a", finding "" as it matches the pattern, and the `search_all` function then detects that the end of this search is at the same point as the end of the previous one, and then stops.

In my example it instead finds "", and then stops by the same reason.

While it can be said to be counterintuitive it also is not a very useful search, and there would have to be a decision over how to change this behavior, do we add all the empty cases? Do we add none?

Doubt that any "fix" will be made for this but if it is desired I can write it.

**Comment by thebluetropics** [comment]
Yes, I thought the same that `a*` could even match empty strings and probably did an infinite loop. However, as far as I know, RegEx doesn't make the `a*` match empty string. Though, I can fix it by changing `*` to `+`.

Anyway, the "a captain eats crew" is just my brief example. I tried using `\S*` to search every word (not using `\w`), and it still returns one exact match using search_all().

Well, I just think this is a bug, because it works on the other place. If somehow this is an intended feature of the godot, I might use `+` instead of `*`.

**Comment by AThousandShips** [comment]
Not all regex engines are the same, but I'm wondering, what do you intend to achieve with using `*`, why do you want to match an empty result?

**Comment by thebluetropics** [comment]
Yeah, that's my fault of using `*` instead of `+`. I always thought that `a*` will match if there are `a`. I don't know whether I need to close this issue or not, but I think I found another bug, I'll post in a separate issue.

**Comment by TheSofox** [comment]
I fixed this issue while trying to fix a different bug: https://github.com/godotengine/godot/pull/85783

The short version is that `search_all` wasn't able to handle zero length results very well.

I did a test with the above "a*" and "a captain eats crew" here: https://regex101.com/r/ynUQra/1

According to the website, there are 20 matches. 4 of them matches `a` while the remainder are empty strings. Trying the Godot engine with the above pull request, I was able to get the exact same result: 

![image](https://github.com/godotengine/godot/assets/3089525/3c8fed36-5d14-4c7d-89dc-abf4b5568738)

Switching back to a version of the engine before my branch, I got this, showing that it was indeed my branch that fixes this issue:

![image](https://github.com/godotengine/godot/assets/3089525/baa55a01-9eb7-4742-8eb4-e90955b9c6b8)
